src/containerapp/azext_containerapp/_utils.py

```diff
@@ -12,147 +12,126 @@
-            # logger.warning("Creating '%s' role assignment under scope '%s'", role, scope)
-            # retry till server replication is done
-            for retry_time in range(0, _RETRY_TIMES):
-                try:
-                    _create_role_assignment(cmd.cli_ctx, role, sp_oid, None, scope, resolve_assignee=False,
-                                            assignee_principal_type='ServicePrincipal')
-                    break
-                except Exception as ex:
-                    if retry_time < _RETRY_TIMES and ' does not exist in the directory ' in str(ex):
-                        time.sleep(5)
-                        logger.warning('  Retrying role assignment creation: %s/%s', retry_time + 1,
-                                       _RETRY_TIMES)
-                        continue
-                    if _error_caused_by_role_assignment_exists(ex):
-                        logger.warning('  Role assignment already exists.\n')
-                        break
-
-                    # dump out history for diagnoses
-                    logger.warning('  Role assignment creation failed.\n')
-                    if getattr(ex, 'response', None) is not None:
-                        logger.warning('  role assignment response headers: %s\n',
-                                       ex.response.headers)  # pylint: disable=no-member
-                    raise
-
-    if show_auth_for_sdk:
-        from azure.cli.core._profile import Profile
-        profile = Profile(cli_ctx=cmd.cli_ctx)
-        result = profile.get_sp_auth_info(scopes[0].split('/')[2] if scopes else None,
-                                          app_id, password, cert_file)
-        # sdk-auth file should be in json format all the time, hence the print
-        print(json.dumps(result, indent=2))
-        return
+def _create_application(client, display_name):
+    from azure.cli.command_modules.role.custom import GraphError
+    body = {"displayName": display_name, "keyCredentials": []}
 
-    result = {
-        'appId': app_id,
-        'password': password,
-        'displayName': app_display_name,
-        'tenant': graph_client.config.tenant_id
-    }
-    if cert_file:
-        logger.warning(
-            "Please copy %s to a safe place. When you run 'az login', provide the file path in the --password argument",
-            cert_file)
-        result['fileWithCertAndPrivateKey'] = cert_file
+    try:
+        result = client.application_create(body)
+    except GraphError as ex:
+        if 'insufficient privileges' in str(ex).lower():
+            link = 'https://docs.microsoft.com/azure/azure-resource-manager/resource-group-create-service-principal-portal'  # pylint: disable=line-too-long
+            raise ValidationError("Directory permission is needed for the current user to register the application. "
+                                  "For how to configure, please refer '{}'. Original error: {}".format(link, ex))
+        raise
```

Comment:
The implementation in main Azure CLI is too generic for your use case. I am glad to see your implementation is much simpler.

Reply:
@StrawnSC looks like we have approval but build pipeline is failing  once we have this working - we can get this build deployed & released. Thank you for all the efforts here.